### PR TITLE
WIP: Type system rework

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This is a huge work in progress PR.
It is aimed at rewriting the type system core, it's aim is to:
- make it easier to work with types removing the explicit distinction between `Type` and `LazyType`
- remove the `Decorators` that made the type signatures too complicated
- improve type inference in dealing with self referencing types
- make the codebase easier to approach with documentation and tests
- (hopefully) make it easier to add new types and functionalities; in this regard we should also add a contributing guide to help future TwinLogix developers and outside contributors

### CODE REVIEW

File: packages/advanced-types/src/types/currency.ts [hunk 1]
```diff
@@ -1,8 +1,9 @@
...
+  /^(AD|AE|AF|AG|AI|AL|AM|AO|AQ|AR|AS|AT|AU|AW|AX|AZ|BA|BB|BD|BE|BF|BG|BH|BI|BJ|BL|BM|BN|BO|BQ|BR|BS|BT|BV|BW|BY|BZ|CA|CC|CD|CF|CG|CH|CI|CK|CL|CM|CN|CO|CR|CU|CV|CW|CX|CY|CZ|DE|DJ|DK|DM|DO|DZ|EC|EE|EG|EH|ER|ES|ET|FI|FJ|FK|FM|FO|FR|GA|GB|GD|GE|GF|GG|GH|GI|GL|GM|GN|GP|GQ|GR|GS|GT|GU|GW|GY|HK|HM|HN|HR|HT|HU|ID|IE|IL|IM|IN|IO|IQ|IR|IS|IT|JE|JM|JO|JP|KE|KG|KH|KI|KM|KN|KP|KR|KW|KY|KZ|LA|LB|LC|LI|LK|LR|LS|LT|LU|LV|LY|MA|MC|MD|ME|MF|MG|MH|MK|ML|MM|MN|MO|MP|MQ|MR|MS|MT|MU|MV|MW|MX|MY|MZ|NA|NC|NE|NF|NG|NI|NL|NO|NP|NR|NU|NZ|OM|PA|PE|PF|PG|PH|PK|PL|PM|PN|PR|PS|PT|PW|PY|QA|RE|RO|RS|RU|RW|SA|SB|SC|SD|SE|SG|SH|SI|SJ|SK|SL|SM|SN|SO|SR|SS|ST|SV|SX|SY|SZ|TC|TD|TF|TG|TH|TJ|TK|TL|TM|TN|TO|TR|TT|TV|TW|TZ|UA|UG|UM|US|UY|UZ|VA|VC|VE|VG|VI|VN|VU|WF|WS|YE|YT|ZA|ZM|ZW)$/
```
Comment (edobrb) [inline packages/advanced-types/src/types/currency.ts:5]: typo from ctrl+c/ctrl+v? (those are country codes)

Comment (giacomocavalieri) [reply packages/advanced-types/src/types/currency.ts:5]: Nice catch! Thanks!

File: packages/model/src/custom/datetime.ts [hunk 2]
```diff
@@ -1,38 +1,62 @@
...
-  return custom(
-    {
-      name: 'datetime',
-      format: 'date-time',
-      encodedType: string(),
-      decode: (input, options, decodeOptions) => {
-        let time: number = Date.parse(input)
-        if (Number.isNaN(time)) {
-          time = Number(input)
-          if (!decodeOptions?.cast || Number.isNaN(time) || time > 864000000000000 || time < -864000000000000) {
```
Comment (edobrb) [inline packages/model/src/custom/datetime.ts:14]: why this check is no more needed? https://www.typescriptlang.org/play?#code/MYewdgziA2CmB00QHMAUZYHcAEARAhgC6yoAcAbACwAMtd9dAlIwLABQokMCSaGOBYmSoNR1AIzMgA

also this old check is wrong (it needs one more zero)

File: packages/model/src/type-system.ts [hunk 3]
```diff
@@ -1,652 +1,1191 @@
...
+  setOptions(options: NumberTypeOptions): NumberType
+  updateOptions(options: NumberTypeOptions): NumberType
+  setName(name: string): NumberType
+}
+
+/**
+ * The options that can be used to define a {@link NumberType `NumberType`}.
+ */
+export type NumberTypeOptions = BaseOptions & {
+  readonly maximum?: [number, 'inclusive' | 'exclusive']
```
[the remark this answers is not in the document]
Comment (edobrb) [reply packages/model/src/type-system.ts:169]: + usually only `minimum` is used and having to write minimum: `[X, 'inclusive']` instead of `minimum: X` is a little bit verbose

File: packages/model/src/custom/datetime.ts (hunk 2 above)
Comment (giacomocavalieri) [reply packages/model/src/custom/datetime.ts:14]: It's because the date constructor already does the work of validating it for us: if the date is invalid the `.valueOf` the built date is a `NaN`
If you check the `tryMakeDate` function you'll see that it does return an error with an invalid date: [playground](https://www.typescriptlang.org/play?ssl=6&ssc=43&pln=1&pc=1#code/GYVwdgxgLglg9mABFATgTwLIEMDWBTAESyjwAoA3LAGxDwC5EwQBbAIzxUQB9EBnVGGADmASgZES3PgOGIA3gFgAUIkQQE-RABNieRAF5GeAO6IJZSjTwjlqlHiggUSAHIt2KAHQxeLrC9IdEk9LWgB5YFIREUQAfkQAIjA4KEQsbV0ExAYgvGUAX2VldTBeOCo8Tyo4IVJUTFxCXVIADgA2ABYABh7evp6ARmigA)

File: packages/model/src/type-system.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (edobrb) [reply packages/model/src/type-system.ts:169]: the first one is X >= 1 & X > 3, so it's X > 3
the second one is X >= 3 & X > 1, so it's X >= 3

File: packages/model/src/custom/datetime.ts (hunk 2 above)
Comment (edobrb) [reply packages/model/src/custom/datetime.ts:14]: ok cool!

File: packages/model/src/type-system.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (edobrb) [reply packages/model/src/type-system.ts:169]: Basically I would execute the check for every options, if one of them fails then return an error.

[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply packages/model/src/type-system.ts:169]: Ok I'll fix this as soon as I can 👍

File: packages/advanced-types/tests/types/property-helper.ts [hunk 4]
```diff
@@ -118,6 +118,16 @@ export function testTypeEncodingAndDecoding<T extends m.Type>(
...
+ *
+ * It's useful for testing all those types whose encode/decode process is not bijective.
+ *
+ * @param type the type to test
+ * @param validValues an array valid raw, decoded and encoded sequence
+ * @param knownInvalidValues an array of invalid values that does not pass the decode step
+ */
 export function testTypeDecodingAndEncoding<T extends m.Type>(
   type: T,
   {
```
Comment (giacomocavalieri) [inline packages/advanced-types/tests/types/property-helper.ts]: Why is this needed when you could call `testTypeEncodingAndDecoding` with `skipInverseCheck` set to true?

File: packages/model/src/result.ts [hunk 5]
```diff
@@ -38,21 +36,20 @@ class Failure<A, E> {
...
-  lazyUnwrap = (fallback: () => A): A => fallback()
+  recover = (fromError: (error: E) => A): A => fromError(this.error)
   or = (other: Result<A, E>): Result<A, E> => other
-  lazyOr = (other: () => Result<A, E>): Result<A, E> => other()
+  lazyOr = (other: (error: E) => Result<A, E>): Result<A, E> => other(this.error)
   match = <B>(_onOk: (value: A) => B, onFailure: (error: E) => B): B => onFailure(this.error)
 }
 
-export function ok<A, E>(value: A): Result<A, E> {
+export function ok<A, E>(value: A): Ok<A, E> {
```
Comment (giacomocavalieri) [inline packages/model/src/result.ts:49]: @edobrb why make this return the more specific value? Does this have any advantage?

Comment (edobrb) [reply packages/model/src/result.ts:49]: I like to return the narrowest type. I would expect to be able to write this:

```typescript
const result = ok(10)
console.log(result.value) //already narrowed
```

instead of:

```typescript
const result = ok(10)
if(result.isOk) {
  console.log(result.value)
}
```

File: packages/model/src/projection.ts [hunk 6]
```diff
@@ -1,9 +1,321 @@
...
+import { filterMapObject, failWithInternalError } from './utils'
 
-export type GenericProjection = true | { [K in string]?: true | GenericProjection }
+/**
+ * This is the Mondrian type describing the structure of a projection: it is either the literal value
+ * `true` or a union composed of two variants: an `all` variant (which is again the literal `true`) and
+ * a `partial` variant which describes a subprojection with possibly many fields; this is why it is described
+ * by an `ObjectType` whose fields can themselves only be valid projections.
+ */
+type Projection =
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Not needed

File: packages/model/src/projection.ts [hunk 7]
```diff
@@ -1,9 +1,321 @@
...
+ * You can read {@link here TODO:} to learn more about what projections are and how they can be used.
+ *
+ * @example ```ts
+ *          const model = types.object({ field1: number, field2: string })
+ *          type Projection = projection.Infer<typeof model>
+ *          // -> true | { field1?: true, field2?: true }
+ *          ```
+ */
+// prettier-ignore
+export type Infer<T extends types.Type> = true | (
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed as is

File: packages/model/src/projection.ts [hunk 8]
```diff
@@ -1,9 +1,321 @@
...
+ * Given a Mondrian {@link Type type} and a projection P, returns the projection of T under P.
+ * You can read {@link here TODO:} to learn more about what projections are and how they can be used.
+ *
+ * @example ```ts
+ *          const model = types.object({ field1: number, field2: string })
+ *          type Projected = projection.Project<typeof model, { field1: true }>
+ *          // -> { readonly field1: true }
+ *          ```
+ */
+export type Project<T extends types.Type, P extends Infer<T>> = types.Infer<projection.ProjectedType<T, P>>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed but also need to exclude Relations in some cases (see tests)

File: packages/model/src/projection.ts [hunk 9]
```diff
@@ -1,9 +1,321 @@
...
+
+// TODO: add doc and instead of types.Infer it should be InferPartialDeep! For now this is just a mock
+//       in order not to stop development of other modules
+// TODO: this should return a result with the missing fields in the error case
+// tells if value is valid (meaning that it only has the fields allowed by the actual projection)
+export function respectsProjection<T extends types.Type>(
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed, returns `Result<true, SomeError>`

File: packages/model/src/projection.ts [hunk 10]
```diff
@@ -1,9 +1,321 @@
...
+  projection: projection.Infer<T>,
+  value: types.Infer<T>,
+): boolean {
+  failWithInternalError('TODO: This is not implemented, if you catch this means we forgot an implementation!')
+}
+
+/**
+ * A record of {@link Projection `Projection`s}.
+ */
+type Projections = Record<string, types.OptionalType<Projection>>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Not needed

File: packages/model/src/projection.ts [hunk 11]
```diff
@@ -1,9 +1,321 @@
...
+ *
+ * @example ```ts
+ *          const model = object({ field1: number, field2: string })
+ *          type Projection = InferProjection<typeof model>
+ *          Infer<Projection>
+ *          // -> true | { field1?: true, field2?: true }
+ *          ```
+ */
+// prettier-ignore
+export type FromType<T extends types.Type>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Cool but i think the end user will never use it. Not needed

File: packages/model/src/projection.ts [hunk 12]
```diff
@@ -1,9 +1,321 @@
...
+ *          // -> union({
+ *          //   all: literal(true)
+ *          //   partial: object({
+ *          //     field1: literal(true).optional()
+ *          //     field2: literal(true).optional()
+ *          //   })
+ *          // })
+ *          ```
+ */
+export function fromType<T extends types.Type>(type: T): projection.FromType<T> {
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Maybe we'll use it in openapi generation. The return type can be more generic like:
```typescript
type UnknownProjection = true | { [K in string]?: true | UnknownProjection }
```

File: packages/model/src/projection.ts [hunk 13]
```diff
@@ -1,9 +1,321 @@
...
+ *          ```
+ * @example ```ts
+ *          const model = object({ field1: number, field2: object({ inner1: string }) })
+ *          type Projection = InferProjection<typeof model>
+ *          ProjectionKeys<Projection>
+ *          // -> "field1" | "field2"
+ *          ```
+ */
+// prettier-ignore
+export type ProjectionKeys<P extends Projection>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: I think the end user will not directly use it.

File: packages/model/src/projection.ts [hunk 14]
```diff
@@ -1,9 +1,321 @@
...
+ * returns the subprojection corresponding to that key.
+ *
+ * @example ```ts
+ *          type Object = ObjectType<{ field1: NumberType, field2: NumberType }>
+ *          SubProjection<ObjectType, "field1">
+ *          // -> true
+ *          ```
+ */
+// prettier-ignore
+export type SubProjection<P extends Projection, K extends ProjectionKeys<P>>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed 
plus if it's deep

File: packages/model/src/projection.ts [hunk 15]
```diff
@@ -1,9 +1,321 @@
...
+ *          subProjection(model, {}, "field")
+ *          // -> undefined
+ *          ```
+ * @example ```ts
+ *          const model = object({ field: number })
+ *          subProjection(model, {}, true)
+ *          // -> true
+ *          ```
+ */
+export function subProjection<const P extends Projection, K extends ProjectionKeys<P>>(
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed
plus is it's deep (dot notation)

File: packages/model/src/projection.ts [hunk 16]
```diff
@@ -1,9 +1,321 @@
...
+ *          projectionDepth<typeof number>(true)
+ *          // -> 0
+ *          ```
+ * @example ```ts
+ *          const model = object({ field1: object({ inner1: number }), field2: number })
+ *          projectionDepth<typeof model>({ field1: { inner1: true } })
+ *          // -> 2
+ *          ```
+ */
+export function depth<P extends Projection>(projection: P): number {
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Needed but it should work with an instance of a projection and not a ProjectionType (copy from branch `develop`)

File: packages/model/src/projection.ts [hunk 17]
```diff
@@ -1,9 +1,321 @@
...
+    const depths = innerProjections.map((inner) => depth(inner.wrappedType))
+    return Math.max(-1, ...depths) + 1
+  }
+}
+
+/**
+ * TODO: add doc
+ */
+// prettier-ignore
+export type ProjectedType<T extends types.Type, P extends projection.Infer<T>>
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Not needed anymore because of projection.respectProjection

File: packages/model/src/projection.ts [hunk 18]
```diff
@@ -1,9 +1,321 @@
...
+  : [T] extends [types.UnionType<infer Ts>] ? [keyof P] extends [keyof Ts]
+    ? types.UnionType<{ [K in keyof P]: P extends Record<K, projection.Infer<Ts[K]>> ? ProjectedType<Ts[K], P[K]> : never }> : never
+  : [T] extends [types.ObjectType<infer _, infer Ts>] ? [keyof P] extends [keyof Ts]
+    ? types.ObjectType<"immutable", { [K in keyof P]: P extends Record<K, projection.Infer<Ts[K]>> ? ProjectedType<Ts[K], P[K]> : never }> : never
+  : never
+
+/**
+ * TODO: add doc
+ */
+export function projectedType<T extends types.Type, P extends projection.Infer<T>>(
```
Comment (edobrb) [inline packages/model/src/projection.ts]: Not needed anymore because of projection.respectProjection

**Review by edobrb — COMMENTED** [review]
I briefly summarized the necessary changes to projection.ts. We also need a function that validates a `PartialDeep<Infer<T>>` that we'll use in combination to projection.respectsProjection.

**Comment by edobrb** [comment]
Excellent work @giacomocavalieri! Thank you!
